core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/batchutilities/BatchingGroupMap.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.batchutilities;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+@SdkInternalApi
+public class BatchingGroupMap<RequestT, ResponseT> {
+    private final Map<String, BatchContext<RequestT, ResponseT>> idToBatchContext;
+    private final AtomicInteger numRequests;
```

Comment:
Why do we need `numRequests`?

Reply:
Since batchingGroupMap contains both requests and responses, the batchingGroupMap can't actually remove a request without removing a response. So in order to check if the number of requests is less than maxBatchItems, you  can't just check the size of the map and instead have to keep track of the number of requests separately.

Reply:
Backlog Item: Address in a separate PR.